### ISSUE TITLE
feat(agent): implement end-to-end Task 1 Q-learning coin collector

### ISSUE THREAD

**Issue by Waffelmanufaktur** [1BlauNitrox/mle-final-project#26]
## Objective

Implement the first complete team-owned learned agent: a self-contained, feature-based tabular Q-learning coin collector for Task 1 in the `coin-heaven` scenario.

The result of this issue must be a genuinely learning, trainable, persistable, and evaluable agent. It must replace the current scaffold-only approach with one vertical implementation that is ready for controlled Task 1 experiments.

## Context

This issue builds on:

- the normative Task 1 contract from #21 and `docs/0007-task-1-baseline-contract.md`;
- the experiment metrics and plotting pipeline from #25; and
- the package boundary proposed in #23.

The implementation should absorb the useful scope of #23 rather than creating a second scaffold followed by more module-level issues. Issue #22 may provide useful lifecycle documentation, but it is not a blocker: callback behavior must be verified directly against the framework and captured in tests where it affects correctness.

## Agent package

Create or complete `agent_code/baseline_agent/` as a self-contained learned agent. Use modules where they clarify responsibilities, but do not create empty placeholders merely to satisfy a directory diagram.

The package must contain at least:

- `callbacks.py` for setup and evaluation-time action selection;
- `train.py` for training callbacks and Q-learning updates;
- a compact feature/state encoder;
- the Q-table and epsilon-greedy policy;
- configurable rewards and hyperparameters;
- relative-path model persistence; and
- an agent README describing the implemented design and known limitations.

### State representation

Implement a compact, hashable state representation suitable for tabular Q-learning. It should include enough information to learn visible-coin navigation, for example:

- whether adjacent movement directions are free or blocked;
- coarse relative direction to the nearest visible coin;
- an optional compact distance bin or equivalent progress signal.

The representation must be documented with its exact values and state-space implications. It must not encode a deterministic pathfinding decision or a rule-based “best action”.

### Action contract

The ordered Task 1 action set is:

```text
UP, RIGHT, DOWN, LEFT, WAIT
```

`BOMB` is excluded during both training and evaluation and must never be selected.

### Learning

Implement standard tabular Q-learning with:

- configurable learning rate `alpha` and discount factor `gamma`;
- zero or explicitly documented initialization for unseen states;
- epsilon-greedy exploration during training;
- a configurable, episode-based epsilon decay with a minimum value;
- seeded random action selection and tie-breaking;
- ordinary transition updates; and
- terminal updates without bootstrapping from a future state.

The initial reward mapping must be minimal, configurable, and documented. It must reward coin collection and discourage invalid actions and unproductive waiting. Optional distance-based reward shaping may be implemented behind a configuration switch, but choosing between minimal and shaped rewards belongs to a follow-up controlled experiment.

### Training and evaluation separation

Training mode must:

- update the Q-table only from training callbacks;
- decay exploration at the documented lifecycle point;
- save a resumable model artifact; and
- expose the optional learning metrics supported by #25, including at least cumulative shaped reward, epsilon, and Q-table size when available.

Evaluation mode must:

- load the model relative to the agent module;
- disable exploration and Q-value updates;
- never write or replace the model artifact;
- use deterministic tie-breaking or a recorded agent seed; and
- work when only `agent_code/baseline_agent/` is copied into a clean framework.

## Experiment-pipeline integration

Use the environment from #25 for the implementation smoke runs. The produced metrics must be sufficient to inspect at least:

- coins and score over episodes;
- episode length or steps per coin;
- invalid-action rate and action distribution;
- cumulative shaped reward;
- epsilon; and
- Q-table growth.

This issue verifies that the agent and data pipeline work together. It does not perform the final held-out evaluation or claim that the completion thresholds in `docs/0007-task-1-baseline-contract.md` have been met.

## Acceptance criteria

- [ ] `agent_code/baseline_agent/` contains a complete, self-contained learned Task 1 agent rather than a scaffold.
- [ ] All five required framework callbacks exist with compatible signatures and handle initial and terminal lifecycle cases correctly.
- [ ] The documented state representation is compact, hashable, and does not encode a rule-based optimal action.
- [ ] Training uses epsilon-greedy tabular Q-learning with correct ordinary and terminal update rules.
- [ ] The agent returns only `UP`, `RIGHT`, `DOWN`, `LEFT`, or `WAIT`; tests confirm that `BOMB` is never selected.
- [ ] Hyperparameters, rewards, exploration schedule, action order, and random seeds are centralized and documented.
- [ ] Save/load round-trips preserve the learned Q-table and all evaluation-relevant configuration.
- [ ] Evaluation is deterministic for a fixed model, world seed, agent seed, framework revision, and configuration.
- [ ] Evaluation performs no exploration, learning, multiprocessing, or model writes.
- [ ] Unit tests cover feature encoding, unseen states, epsilon-greedy behavior, seeded tie-breaking, ordinary Q-updates, terminal Q-updates, action validity, and persistence.
- [ ] A short headless `coin-heaven` training smoke run completes and produces a non-empty learned model plus metrics through #25.
- [ ] A subsequent evaluation smoke run loads that artifact, completes without mutation, and produces the expected evaluation metrics and plots.
- [ ] The agent's README documents hypothesis, model, state representation, rewards, exploration, training command, artifact path, limitations, and the fact that scientific performance evaluation remains follow-up work.
- [ ] Smoke-test-only models, raw logs, caches, and temporary figures are removed or remain in gitignored output locations.
- [ ] Relevant lint, unit, compile, training-smoke, and evaluation-smoke checks pass.

## Out of scope

- Final Task 1 hyperparameter optimization.
- Selecting minimal versus shaped rewards based on performance.
- Running the five-training-seed and 100-held-out-world-seed completion evaluation.
- Claiming that the Task 1 completion thresholds have been achieved.
- Bombs, crates, opponents, deep learning, self-play, or tournament optimization.
- Building additional experiment infrastructure beyond what #25 provides.

## Dependencies and issue relationship

- Uses the completed Task 1 contract from #21.
- Depends on #25 for final smoke-run metrics and plots; agent implementation may proceed in parallel, but this issue is not complete until integration works.
- Supersedes the scaffold-only implementation scope of #23. Once this issue is accepted as the implementation plan, #23 should be closed as superseded or rewritten to avoid duplicate work.
- #22 is informative but not a blocker.

## Follow-up

Create one controlled Task 1 experiment issue after this implementation is complete. That experiment should compare the minimal baseline against a single justified feature or reward-shaping variant using the protocol and held-out seed rules already defined in the repository.
